source/simulator/assemblers/advection.cc

```diff
@@ -468,6 +468,307 @@ namespace aspect
+                += tau *
+                   (
+                     (current_u_f) *
+                     scratch.grad_phi_field[i] *
+                     (
+                       field_term_for_rhs
+                       +
+                       reaction_term
+                     )
+                   ) * JxW;
+
+
+              for (unsigned int j=0; j<advection_dofs_per_cell; ++j)
+                {
+                  data.local_matrix(i,j)
+                  += (
+                       (time_step * scratch.artificial_viscosity
+                        * (scratch.grad_phi_field[i] * scratch.grad_phi_field[j]))
+                       + ((time_step * (scratch.phi_field[i] * (current_u_f * scratch.grad_phi_field[j])))
+                          + (bdf2_factor * scratch.phi_field[i] * scratch.phi_field[j]))
+                     )
+                     * JxW;
+                }
+
+              if (use_supg)
+                {
+                  for (unsigned int j=0; j<advection_dofs_per_cell; ++j)
+                    {
+                      // Note that we assume that the conductivity is constant, otherwise we would need to
+                      // compute div (kappa grad T), which we don't have access to.
+                      data.local_matrix(i,j)
+                      += tau *
+                         (
+                           (current_u_f) *
+                           scratch.grad_phi_field[i] *
+                           (
+                             -time_step * conductivity * scratch.laplacian_phi_field[j]
+                             +
+                             (
+                               (time_step * current_u_f * scratch.grad_phi_field[j])
+                               +
+                               (bdf2_factor * scratch.phi_field[j])
+                             )
+                           )
+                         ) * JxW;
+                    }
+                }
+            }
+        }
+    }
+
+
+
+    template <int dim>
+    std::vector<double>
+    DarcySystem<dim>::compute_residual(internal::Assembly::Scratch::ScratchBase<dim> &scratch_base) const
+    {
+      internal::Assembly::Scratch::AdvectionSystem<dim> &scratch = dynamic_cast<internal::Assembly::Scratch::AdvectionSystem<dim>& > (scratch_base);
+      const typename Simulator<dim>::AdvectionField advection_field = *scratch.advection_field;
+      const unsigned int n_q_points = scratch.finite_element_values.n_quadrature_points;
+      std::vector<double> residuals(n_q_points);
+      if (advection_field.is_temperature())
+          {
+            return 0.;
+          }  
+
+
+      const MaterialModel::MeltOutputs<dim> *melt_outputs = scratch.material_model_outputs.template get_additional_output<MaterialModel::MeltOutputs<dim>>();
+      for (unsigned int q=0; q < n_q_points; ++q)
+        {
+
+          const double K_D = melt_outputs->permeabilities[q] / melt_outputs->fluid_viscosities[q];
```

Comment:
Eventually add documentation here about these variables, which are the key component of the of the new advection method.